### PR TITLE
[K8s] Custom Image Support for Kubernetes Instances

### LINKED ISSUE #2599: [k8s] Custom image support for Kubernetes
We need to support custom images for Kubernetes clusters, possibly set via `image_id: docker:<image_id>`. 

We'd need to take care of a few things/refactor our k8s provisioning:

1. Install and configure openssh as a part of setup to ensure ssh access
2. Make sure basic dependencies are installed, and install them if necessary/add notes to docs about which dependencies are required

### PR DESCRIPTION
**Pull request by landscapepainter**
This PR resolves #2599 

When users opt to provide their custom images for setting up the k8s instance, instead of relying on Skypilot's default image built with `Dockerfile_k8s`, certain assumptions and adjustments are made to ensure smooth operation:

**Base Distribution:** We expect users to provide images that are based on Debian-based distributions.
**User Privilege:** It's crucial for users to configure the default user's privilege correctly within their custom image by either setting the default user to have root privilege or installing `sudo`.
**Dependency Installation & SSH Setup:** Instead of baking these directly into the image, we've shifted the responsibility of installing dependencies and setting up SSH to the `ray` manifest. These processes are triggered after the pods are active and are managed within `node_provider.py` under the `create_node()` method.
**SSH user update:** We have default `ssh_user` of `sky` set in `kubernetes-ray.yml.j2`. However, as users' may have different user name, we need to update the `ssh_user` in the cluster yaml and the `SSHCommandRunner` used within NodeProvider. This is each done from `node_provider.py/create_node()` and `node_provider.py/get_command_runner()`.

By adopting these measures, we aim to offer greater flexibility for users bringing their own images while ensuring consistent system behavior.

Update: This PR supports public registries only, and private registries may be supported in near future. Until then, there will be a doc on how to setup for the private registry by the user.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

Tested Base Images:
- [x] continuumio/miniconda3:22.11.1
- [x] nvidia/cuda:12.2.0-devel-ubuntu20.04
- [x] pytorch/pytorch:latest
- [x] pytorch/pytorch:2.1.0-cuda12.1-cudnn8-runtime
- [x] tensorflow/tensorflow:latest
- [x] tensorflow/tensorflow:nightly-gpu-jupyter
- [x] ubuntu:latest
- [x] gpuci/miniforge-cuda:11.4-devel-ubuntu18.04

Tested Registries:
- [x] AWS ECR public
- [x] GCP Artifact Registry public
- [x] Docker Registry public

Early failovers:
- [x] AWS ECR private handled with error message
- [x] Docker private registry handled with error message with early failure
- [x] GCP Registry failover when user's not configured for the GCP account
- [x] GCP Registry failover when user's configured GCP account does not have IAM authorization
- [x] Attempting to pull from non-existent repository
- [x] Attempting to pull from non-existent tag name with correct repository

### CODE REVIEW

**Comment by landscapepainter** [comment]
Adding `curl` to be installed in `setup_commands` at `kubernetes-ray.yaml.j2` to address #2799.

**Comment by landscapepainter** [comment]
Additional tests for early failover:
- [x] AWS ECR private handled with error message
- [x] Docker private registry handled with error message with early failure
- [x] GCP Registry failover when user's not configured for the GCP account
- [x] GCP Registry failover when user's configured GCP account does not have IAM authorization
- [x] Attempting to pull from non-existent repository
- [x] Attempting to pull from non-existent tag name with correct repository

**Comment by romilbhardwaj** [comment]
Thanks @landscapepainter! Excited for this feature.

I was trying it out with the following YAML:

```
resources:
  image_id: docker:ubuntu:focal

setup: |
  echo setup

run: |
  echo hi
```

But it seems to be getting stuck on getting SSH access (see provision logs):
```
...
2023-12-29 11:06:07,121	VINFO command_runner.py:365 -- Running `uptime`
2023-12-29 11:06:07,121	VVINFO command_runner.py:367 -- Full command is `ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_8d2ba49c09/2be00d9d04/%C -o ControlPersist=10s -o ProxyCommand=ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  -o Port=22 -o ConnectTimeout=10s sky@10.244.0.5 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (uptime)'`
kex_exchange_identification: Connection closed by remote host
Connection closed by UNKNOWN port 65535
2023-12-29 11:06:08,644	INFO updater.py:312 -- SSH still not available (SSH command failed.), retrying in 5 seconds.
...
```

Looks like the pod is running successfully with the custom image:
```
(base) ➜  ~ kubectl describe pod custom-2ea4-ray-head | grep image
  Normal  Pulling    7m7s  kubelet            Pulling image "ubuntu:focal"
  Normal  Pulled     7m3s  kubelet            Successfully pulled image "ubuntu:focal" in 3.615649459s
``` 

Can we add additional logging for `run_command_on_pods` (perhaps with logging.debug level) to know if the initialization commands are successful?

**Comment by romilbhardwaj** [comment]
Made a few fixes and added `pod_override` ability to `~/sky/config.yaml`. This allows users to specify custom pod spec overrides, such as custom labels, imagePullSecrets, imagePullPolicy etc as required through the sky config yaml. Helps a lot by allowing users to configure pods to their environments. 

**Quick guide** for anyone interested in trying out:
To use a custom image, use the `image_id: docker:<your-image-tag>` field under resources field in your YAML. Here's an example:
```
resources:
  cloud: kubernetes
  accelerators: T4:1
  image_id: docker:638892464027.dkr.ecr.us-west-2.amazonaws.com/romilpvtrepo:latest # Or nvidia/cuda:11.8.0-devel-ubuntu18.04
```

If your custom image is hosted on a private registry, you'll need to setup the Kubernetes secret to access it. Once the secret is setup, add the following to your `~/.sky/config.yaml`:
```
kubernetes:
  pod_override:
    spec:
      imagePullSecrets:
        - name: <your-secret-name>
```
This will automatically add the secret to the pod spec when SkyPilot creates the pod.

File: sky/templates/kubernetes-ray.yml.j2 [hunk 1]
```diff
@@ -288,17 +280,19 @@ setup_commands:
...
-    mkdir -p ~/sky_workdir && mkdir -p ~/.sky/sky_app && touch ~/.sudo_as_admin_successful;
+    mkdir -p ~/sky_workdir && mkdir -p ~/.sky/sky_app && sudo touch ~/.sudo_as_admin_successful;
     (pip3 list | grep skypilot && [ "$(cat {{sky_remote_path}}/current_sky_wheel_hash)" == "{{sky_wheel_hash}}" ]) || (pip3 uninstall skypilot -y; pip3 install "$(echo {{sky_remote_path}}/{{sky_wheel_hash}}/skypilot-{{sky_version}}*.whl)[remote]" && echo "{{sky_wheel_hash}}" > {{sky_remote_path}}/current_sky_wheel_hash || exit 1);
     sudo bash -c 'rm -rf /etc/security/limits.d; echo "* soft nofile 1048576" >> /etc/security/limits.conf; echo "* hard nofile 1048576" >> /etc/security/limits.conf';
     sudo grep -e '^DefaultTasksMax' /etc/systemd/system.conf || (sudo bash -c 'echo "DefaultTasksMax=infinity" >> /etc/systemd/system.conf'); sudo systemctl set-property user-$(id -u $(whoami)).slice TasksMax=infinity; sudo systemctl daemon-reload;
     mkdir -p ~/.ssh; (grep -Pzo -q "Host \*\n  StrictHostKeyChecking no" ~/.ssh/config) || printf "Host *\n  StrictHostKeyChecking no\n" >> ~/.ssh/config;
+    pip3 install ray==2.4.0 aiohttp opencensus prometheus_client aiohttp_cors kubernetes==28.1.0;
```
Comment (Michaelvll) [inline sky/templates/kubernetes-ray.yml.j2:295]: Ray installation sould go before skypilot installation and needs some patch to make it work. Reference: https://github.com/skypilot-org/skypilot/blob/efdd07deb7943666d67f000f6e09701e5161ecae/sky/templates/aws-ray.yml.j2#L163

https://github.com/skypilot-org/skypilot/blob/efdd07deb7943666d67f000f6e09701e5161ecae/sky/templates/aws-ray.yml.j2#L170

File: sky/skylet/providers/kubernetes/node_provider.py [hunk 2]
```diff
@@ -363,22 +455,33 @@ def _set_env_vars_in_pods(self, new_nodes):
...
-                tty=False,
-                _request_timeout=kubernetes.API_TIMEOUT)
+            run_command_on_pods(new_node.metadata.name, self.namespace,
+                                set_k8s_env_var_cmd)
+
+    def _update_ssh_user_config(self, new_nodes, cluster_name_with_hash):
+        get_k8s_ssh_user_cmd = ['/bin/sh', '-c', ('echo $(whoami)')]
+        for new_node in new_nodes:
+            ssh_user = run_command_on_pods(new_node.metadata.name,
+                                           self.namespace, get_k8s_ssh_user_cmd)
```
Comment (Michaelvll) [inline sky/skylet/providers/kubernetes/node_provider.py:472]: Don't have to loop through all the nodes? 
```suggestion
        ssh_user = run_command_on_pods(new_nodes[0].metadata.name,
                                           self.namespace, get_k8s_ssh_user_cmd)
```

File: sky/skylet/providers/kubernetes/node_provider.py [hunk 3]
```diff
@@ -423,17 +526,35 @@ def create_node(self, node_config, tags, count):
...
+        self._wait_for_pods_to_run(new_nodes_with_jump_pod)
+        logger.info(config.log_prefix +
+                    f'Checking if user in image has sufficient privileges.')
+        self._check_user_privilege(new_nodes)
+        logger.info(config.log_prefix +
+                    f'Setting up SSH in pod.')
+        self._setup_ssh_in_pods(new_nodes)
+        logger.info(config.log_prefix +
+                    f'Setting up environment variables in pod.')
         self._set_env_vars_in_pods(new_nodes)
```
Comment (Michaelvll) [inline sky/skylet/providers/kubernetes/node_provider.py:553]: Do we need to setup envs for jump pod?

File: sky/skylet/providers/kubernetes/node_provider.py [hunk 4]
```diff
@@ -345,6 +382,61 @@ def _wait_for_pods_to_run(self, new_nodes):
...
+             'export DEBIAN_FRONTEND=noninteractive;'
+             '$(prefix_cmd) apt-get update;'
+             '$(prefix_cmd) apt install openssh-server rsync -y; '
+             '$(prefix_cmd) mkdir -p /var/run/sshd; '
+             '$(prefix_cmd) sed -i "s/PermitRootLogin prohibit-password/PermitRootLogin yes/" /etc/ssh/sshd_config; '
+             '$(prefix_cmd) sed "s@session\\s*required\\s*pam_loginuid.so@session optional pam_loginuid.so@g" -i /etc/pam.d/sshd; '
+             'cd /etc/ssh/ && $(prefix_cmd) ssh-keygen -A; '
+             '$(prefix_cmd) mkdir -p ~/.ssh; '
+             '$(prefix_cmd) cp /etc/secret-volume/ssh-publickey ~/.ssh/authorized_keys; '
+             '$(prefix_cmd) service ssh restart')
```
Comment (Michaelvll) [inline sky/skylet/providers/kubernetes/node_provider.py:431]: Do we need the following?
```suggestion
             '$(prefix_cmd) chmod 600 ~/.ssh/authorized_keys; '
             '$(prefix_cmd) service ssh restart')
```

Comment (Michaelvll) [inline sky/skylet/providers/kubernetes/node_provider.py:431]: We probably need this:
```suggestion
            '$(prefix_cmd) chwon -R $(whoami) ~/.ssh;'
             '$(prefix_cmd) service ssh restart')
```

File: sky/skylet/providers/kubernetes/node_provider.py (hunk 3 above)
Comment (landscapepainter) [reply sky/skylet/providers/kubernetes/node_provider.py:553]: The env vars are necessary for pods to interact with GPUs, so it's only necessary for the instance pods, not jump pods.

**Comment by Michaelvll** [comment]
Thanks @landscapepainter for submitting this PR! Since the changes has been adopted into #3019 and merged into master, probably we should close this PR. : )

**Review by Michaelvll — APPROVED** [review]
Thanks for adding the support for k8s custom image @landscapepainter! Most of the parts of this PR has been adopted in #3019, but we should have got this PR in first before that PR to avoid losing the commit history. Let's get this PR in to keep the commit history.

Note for people looking at this PR: the `node_provider` is deprecated for k8s support and all the awesome changes/commits from @landscapepainter  has been applied to #3019. We merged this PR for tracking the commit history. : )
